### ISSUE TITLE
Performance compared to koala2

### ISSUE THREAD

**Issue by done712** [bradbase/xlcalculator#17]
Based on some tests that I ran xlcalculator seems to be much slower than koala2 (https://github.com/vallettea/koala/). However as xlcalculator can run with newer versions of openpyxl I would rather prefer using xlcalculator. Is the performance issue something that you are aware of? Or am I the only one noticing this? My tests revolve around reading in an Excel model, changing the value of one cell and evaluating the value of a second cell (that depends on the first cell).

**Comment by bradbase** [comment]
Hi @done712 

I welcome your performance test. Thank you for taking the time.

At the moment I am only aware of one other person using xlcalculator and, as far as I'm aware, performance is satisfactory in that use case.

In a very early iteration of xlcalculator I compared it with Koala and PyCel. I have made the tests available https://github.com/bradbase/Perfomance_testing_Python_Excel_calculators. You are welcome to add your test cases to this repo. I set the repo up without expecting others to be involved so I'm happy to help get your test cases in if you would like.

When you say xlcalculator is slow, have you identified what quality is slow? There are many aspects. These include;
- Excel file reading
- Excel file writing
- parsing the Excel file to identify the cells
- parsing the Excel file to identify the formulas
- parsing the formulas into AST
- converting the AST to Python code
- evaluating the Python code

Depending on what you are using xlcalculator for you'll obviously prioritize a couple of these over others as being the most important. I would like to think xlcalculator can become the benchmark in all of them, but it's a young project and things take time.

In my original testing I found one of the most difficult things for me to optimize was what I'm calling "chained cells", cells which rely on previous results before they can be evaluated. But this didn't show as an issue with only a couple of cells. It became an issue with tens and hundreds of cells.

In any case, since the above mentioned testing was done, xlcalculator has had some substantial changes made and I have not yet re-run the tests.

There are a number of places within xlcalculator that could be slow. I made some very poor choices early on which have made parts of the code difficult to understand and very difficult to troubleshoot and, likely, slower than they need to be. I intend to replace those parts. Some of them are named in the TODO list on the code page of this repo.

If we can identify which aspects of performance you value, maybe we can focus some attention on that?

Cheers,

Brad
